Mini halos

qyx268 opened the pull request:
including a second population into 21cmFAST (Qin et al. in prep.)

This feature separates atomic-cooling galaxies (ACGs) and molecular-cooling galaxies (MCGs), and considers feedback for them differently. We now include photoheating and photodissociating feedback, with a few more parameters added for the minihalo population.

steven-murray commented on src/py21cmfast/src/IonisationBox.c:
```diff
@@ -576,44 +886,170 @@ LOG_ULTRA_DEBUG("while loop for until RtoM(R)=%f reaches M_MIN=%f", RtoM(R), M_M
+
                         curr_dens = *((float *)deltax_filtered + HII_R_FFT_INDEX(x,y,z));
+						if (ION_EFF_FACTOR_MINI > 1e-19){
+                        	prev_dens = *((float *)prev_deltax_filtered + HII_R_FFT_INDEX(x,y,z));
```
this looks like bad formatting

steven-murray commented on src/py21cmfast/src/UsefulFunctions.c:
```diff
@@ -763,4 +763,30 @@ void print_corners_real(float *x, int size){
+    return  molecular_cooling_threshold(z) * (1. + 22.8685 * pow(J_21_LW, 0.47));
+}
+
+#define HALO_BIAS     (double) 2.0
```
Can these constants be moved to constants.h?

steven-murray commented:
The test that is failing is the `test_wrapper/test_ib_new_seed` test. The point of the test is to create two boxes with different seeds and ensure that not all of the cells in the xHI box are the same. It _looks_ like in the case of python 3.6, for some reason, all cells are xHI=1 for both seeds. This is at z=15. Perhaps we just need to change the redshift?

qyx268 commented:
there were issues with index, which caused wrong results and blocked the tests. I'm varying parameters and testing the results. 

Pending: better testing for mini-halos.

qyx268 commented:
testing has been done by varying parameters that describe minihalos. The results look reasonable and as expected. I've added a notebook in tutorials to demonstrate the results. 

@steven-murray  do you know what is going on with travis-ci/pr? Why has push passed while the pr failed?

steven-murray commented:
Looks like this is segfaulting on Travis?

qyx268 commented:
I cannot pass the tests! Struggling to reproduce the test data

steven-murray commented:
You should probably merge master in again, as it has fixes for the memory leak, and maybe some other important fixes.

qyx268 commented:
Yes, I just did it. 
I found the differences are not significant, but still larger than atol=1e-5, rtol=1e-4. Do I need to worry that different compilers or versions of libraries might cause different results?

steven-murray commented:
You shouldn't need to worry about that, since Travis should still be using the same compiler/libraries as it was before when it passed. 

Looking at the outputs in the tests, it seems like some k-bins have quite significant differences (30-40%), especially at high k. Have you added any code that changes output when the mini halos are turned off?

qyx268 commented:
okay, finally, I passed all the test!!! 

Note that there is a bug in recombination where the boundary has been set wrongly.

qyx268 commented:
I was told to celebrate this, here is my celebration gift -- a good meme

![image](https://user-images.githubusercontent.com/15994713/69815554-dcaa5980-11f6-11ea-9dbc-18645fd01d81.png)

qyx268 commented:
celebrated too early... 

@steven-murray  can you help me with the failed test from black??

andreimesinger commented:
https://i.kym-cdn.com/photos/images/original/001/001/482/f52.gif <https://i.kym-cdn.com/photos/images/original/001/001/482/f52.gif>


> On 28.11.2019., at 16:18, Yuxiang Qin <notifications@github.com> wrote:
> 
> celebrated too early...
> 
> @steven-murray <https://github.com/steven-murray> can you help me with the failed test from black??
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub <https://github.com/21cmfast/21cmFAST/pull/37?email_source=notifications&email_token=ADH2UATLFXDUHJRMDOW7FV3QV7OOHA5CNFSM4I5FSDKKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFM4KQY#issuecomment-559531331>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADH2UASRVWWUQMJSLO5PE4LQV7OOHANCNFSM4I5FSDKA>.
>

steven-murray commented:
Sorry for the delay -- thanksgiving over here!

Ah yes, `pre-commit` is certainly useful. If you haven't now already, please install it and run `pre-commit install`! Looks like you've got everything passing now though! Well done!

The bug in recombinations -- did you fix that, or are you pointing out that it is still there? If it's still there, can you make an issue for it (I'll approve this PR and we can move forward).

qyx268 commented:
Yes, everything have been fixed including pre-commit and the recombination

steven-murray commented on src/py21cmfast/wrapper.py:
```diff
@@ -232,6 +232,9 @@ class UserParams(StructWithDefaults):
         5: CLASS output
+
+    L_FACTOR : float, optional
+        Factor relating cube length to filter radius = (4PI/3)^(-1/3), default is 0.620350491
```
This docstring is confusing -- how could a constant like this be a parameter?

qyx268 replied:
@andreimesinger can you comment on this? I'm not sure if L_FACTOR should be a variable.

qyx268 replied:
@steven-murray so I need to use L_FACTOR to initialize R_BUBBLE_MIN, which is the reason I've moved L_FACTOR into the wrapper. As @andreimesinger said, it's a constant and shouldn't be put in parameters. Would you prefer to putting it as the number directly in the code (.c and .py) or simply remove the docstring (so still store it in UserParams but don't allow users to change it)?

steven-murray replied:
I think in this case, the easiest thing is just to set `L_FACTOR = (4*np.pi/3)**(1./3)` in the `R_BUBBLE_MIN` property, since it will never change. It's slight duplication (i.e. it's defined in C and Python), but it's very small and much simpler this way.

qyx268 replied:
please see if the current way makes sense. I would do pr_feature now.

andreimesinger commented:
no, it's a geometric constant

> On 3 Dec 2019, at 09:55, Yuxiang Qin <notifications@github.com> wrote:
> 
> ﻿
> @qyx268 commented on this pull request.
> 
> In src/py21cmfast/wrapper.py:
> 
> > @@ -232,6 +232,9 @@ class UserParams(StructWithDefaults):
>          3: PEEBLES
>          4: WHITE
>          5: CLASS output
> +
> +    L_FACTOR : float, optional
> +        Factor relating cube length to filter radius = (4PI/3)^(-1/3), default is 0.620350491
> @andreimesinger can you comment on this? I'm not sure if L_FACTOR should be a variable.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

qyx268 commented:
The default model when doing pr-feature assumes False USE_MASS_DEPENDENT_ZETA and Flase INHOMO_RECO. However, I need these two flag to be True and set F_STAR7_MINI to be a large value to test the new feature. 

I don’t think the current version of pr-feature works for passing multiple arguments. I would change the default in cli to have USE_MASS_DEPENDENT_ZETA and INHOMO_RECO set to True for this comparison.

steven-murray commented:
Sorry for the delay, @qyx268. I'm currently at a conference and not finding time to get to fix this. I'll flag it as urgent on my to-do list. I'll try changing the `pr-feature` script to accept more than one "change".

On the other hand, perhaps you should change the code so that if you have mini-halos turned on, then `USE_MASS_DEPENDENT_ZETA` and `INHOMO_RECO` get switched on automatically. Can you do that?

qyx268 commented:
Hey @steven-murray , I've thought about it. I don't think automatically setting USE_MASS_DEPENDENT_ZETA and INHOMO_RECO to True when minihalos are included is what we want. 

I'm using the fstar normalization of minihalos (F_STAR7_MINI) as the "switch" for whether minihalos are included or not. When F_STAR7_MINI is very large, minihalo features are switched on. This way, as we discussed before, allows us to not include another flag such as USE_MINIHALO. In fact, USE_MASS_DEPENDENT_ZETA can also be eliminated. We can switch on relevant features by setting the fstar normalization of atomic cooling halos (F_STAR10) to a large value. 

Following the configuration of USE_MASS_DEPENDENT_ZETA and F_STAR10 (i.e. we have to manually set USE_MASS_DEPENDENT_ZETA to True in order to use features that include F_STAR10), I think we should keep the same configuration -- we have to manually set USE_MASS_DEPENDENT_ZETA (and INHOMO_RECO) to True in order to use features that include F_STAR7_MINI.

If we really want to automatically set USE_MASS_DEPENDENT_ZETA and INHOMO_RECO to True when minihalos are included, I would suggest including the USE_MINIHALO flag in FlagOptions, and do the automation when USE_MINIHALO is on.

Do these make sense to you?

qyx268 commented:
@steven-murray also, do you think, for pr-feature, it's better to have the option to set what the default model should be? For instance, the minihalo implementation should be compared to Park+18, not the current default which uses the 3 parameters. Or if someone's development is based on the minihalo branch, then they should compare to the default minihalo model.

steven-murray commented:
I think I understand what you mean, although I'm not *sure* I agree with the conclusion :-). Let me try to restate the options:

Basically, the problem is that some options could be just as easily specified by a flag or a float value. The current example is that `F_STAR7_MINI` is a float value, and it determines whether mini-halos are used, by checking whether its value is > 1e-9. We could alternatively have both this value _and_ a flag, say `USE_MINI_HALOS`: the latter determining whether to do mini-halos, and the former setting the relevant parameter. At this point, we've opted for the former option (just a single float value which turns things on/off and also specifies the actual value to use). However, for other features, we've opted for the latter: including an explicit flag. We should _at least_ make this consistent. There is a sub-option when having both float and flag: the float option sets the default of the flag.

So let's make this more explicit. Let's assume we have a parameter `X` and flag `USE_X` (the former in `AstroParams` and the latter in `FlagOptions`). The options we have are:

1. Do not even have `USE_X`. Instead, let the value of `X` determine whether the feature is actually used. **Pros:** only one point of reference, one less parameter. **Cons:** No explicit flag to check for whether the feature was used (especially if examining the data the next year...), and the condition for using the feature is defined _in the code_ (i.e. there are comparisons like `if X > 1e-9`). Also, depending on the kind of parameter/condition, there may be an arbitrary choice to make... what is the threshold for turning it off? Will this make a quantitative jump if running multiple models that go through the threshold?
2. Use both `X` and `USE_X`, and exclusively use the explicit value of `USE_X` to determine whether the feature is "on". **Pros:** very easy to tell whether the feature was run or not. **Cons:** could be slightly more confusing for the user to set the right parameters. They *may* assume that by setting `USE_X` they will get a reasonable default value of `X`. Conversely, they may assume that by setting `X`, the code will use that value instead of ignoring it. Also, it means they have to set two things instead of one. 
3. Use both `X` and `USE_X`, but if `X` is beyond a given threshold, either *by default* set `USE_X` to True, or even _force_ it to be True (i.e. `USE_X` really becomes a way to quickly tell whether `X` was beyond the threshold, rather than a truly settable parameter. **Pros:** The expected behaviour happens, but with the bonus of having an easy-to-check reference for whether the feature was actually used. **Cons:** the code then has to have `FlagOptions` depend on `AstroParams`, which is a bit awkward, since at the moment, the reverse is true, which could imply a circular dependency. We could try switching this around for existing parameters though.
4. Use both `X` and `USE_X`, but set the default of `X` based on the value of `USE_X`. **Pros: ** the default behaviour only requires setting one parameter, and you have the benefit of the explicit switch. Despite the necessity of one struct depending on another, at least in this instance, we already have `AstroParams` dependent on some parameters from `FlagOptions`, so it's easier to continue this trend. **Cons:** you could have the weird behaviour that you set `X` to a value that clearly should make the feature be used, but it doesn't get used by default. On the other hand, we *could* just issue a warning when this is the case, and let the user deal with it. 

I'm going to bring a few more people into this discussion, because I don't want to make executive decisions at this level of the API. Whatever we do, we should try to make it consistent for all these kinds of parameters.

@BradGreig, @catherinewatkinson and @andreimesinger .... any thoughts? I think I'm leaning towards option 4, with a warning if `X` has a value that looks like it should be used, but `USE_X` is false.

steven-murray commented:
Also, concerning `pr-feature`, I'll have to have a look at how feasible it is to change the default. I don't want it to be too complex. But I'll try to make that happen.

andreimesinger commented:
4 seems most straightforward imo.  the use blah flag should be the determining factor. issue warning about the default choice tou are setting

> On 12 Dec 2019, at 21:10, Steven Murray <notifications@github.com> wrote:
> 
> ﻿
> I think I understand what you mean, although I'm not sure I agree with the conclusion :-). Let me try to restate the options:
> 
> Basically, the problem is that some options could be just as easily specified by a flag or a float value. The current example is that F_STAR7_MINI is a float value, and it determines whether mini-halos are used, by checking whether its value is > 1e-9. We could alternatively have both this value and a flag, say USE_MINI_HALOS: the latter determining whether to do mini-halos, and the former setting the relevant parameter. At this point, we've opted for the former option (just a single float value which turns things on/off and also specifies the actual value to use). However, for other features, we've opted for the latter: including an explicit flag. We should at least make this consistent. There is a sub-option when having both float and flag: the float option sets the default of the flag.
> 
> So let's make this more explicit. Let's assume we have a parameter X and flag USE_X (the former in AstroParams and the latter in FlagOptions). The options we have are:
> 
> Do not even have USE_X. Instead, let the value of X determine whether the feature is actually used. Pros: only one point of reference, one less parameter. Cons: No explicit flag to check for whether the feature was used (especially if examining the data the next year...), and the condition for using the feature is defined in the code (i.e. there are comparisons like if X > 1e-9). Also, depending on the kind of parameter/condition, there may be an arbitrary choice to make... what is the threshold for turning it off? Will this make a quantitative jump if running multiple models that go through the threshold?
> Use both X and USE_X, and exclusively use the explicit value of USE_X to determine whether the feature is "on". Pros: very easy to tell whether the feature was run or not. Cons: could be slightly more confusing for the user to set the right parameters. They may assume that by setting USE_X they will get a reasonable default value of X. Conversely, they may assume that by setting X, the code will use that value instead of ignoring it. Also, it means they have to set two things instead of one.
> Use both X and USE_X, but if X is beyond a given threshold, either by default set USE_X to True, or even force it to be True (i.e. USE_X really becomes a way to quickly tell whether X was beyond the threshold, rather than a truly settable parameter. Pros: The expected behaviour happens, but with the bonus of having an easy-to-check reference for whether the feature was actually used. Cons: the code then has to have FlagOptions depend on AstroParams, which is a bit awkward, since at the moment, the reverse is true, which could imply a circular dependency. We could try switching this around for existing parameters though.
> Use both X and USE_X, but set the default of X based on the value of USE_X. **Pros: ** the default behaviour only requires setting one parameter, and you have the benefit of the explicit switch. Despite the necessity of one struct depending on another, at least in this instance, we already have AstroParams dependent on some parameters from FlagOptions, so it's easier to continue this trend. Cons: you could have the weird behaviour that you set X to a value that clearly should make the feature be used, but it doesn't get used by default. On the other hand, we could just issue a warning when this is the case, and let the user deal with it.
> I'm going to bring a few more people into this discussion, because I don't want to make executive decisions at this level of the API. Whatever we do, we should try to make it consistent for all these kinds of parameters.
> 
> @BradGreig, @catherinewatkinson and @andreimesinger .... any thoughts? I think I'm leaning towards option 4, with a warning if X has a value that looks like it should be used, but USE_X is false.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

catherinewatkinson commented:
I think 4 is reasonable. But I would personally opt for 1. but have a hidden/internal flag for USE_X that is set to True when X is provided in the call and False otherwise. I think this provides optimal simplicity from the users point of view, whilst having the very desirable feature that you can tell if it the feature was on or off in that run.

steven-murray commented:
I think @catherinewatkinson is actually advocating for one of the variants of (3): `USE_X` being a non-settable property of `FlagOptions` which is determined explicitly by the value of `X`. I do really like this mode, but it comes with some technical difficulties (not least of which is that currently `AstroParams` is instead dependent on `FlagOptions`). The other technical difficulty is that these non-settable parameters may behave differently for caching purposes than the settable ones -- eg. they may not map to the C struct. 

Actually, thinking about this as I write... I think (4) really is the way to go. The reason for this is that it's completely reasonable to ask, for example, to do mini-halos, and yet set the F_STAR7_MINI to 1e-10. Even if "nothing happens", it could almost be treated as a test case to set the variable so low and yet still do the process. There is really no absolute natural threshold for any variable, and I doubt we should *enforce* that the process is turned off at any threshold (consider MCMC over the variable, and potential discontinuities that might arise). We can of course issue warnings if the parameter is below the threshold and the flag is still true (to say "hey, things will run faster and be just as accurate if you set `USE_X=False`") and alternatively if the flag is *false* but the value is above the threshold. 

Since @andreimesinger already endorsed this, and @catherinewatkinson doesn't disagree, I'll give it a ~day for @BradGreig to weigh in and/or any other comments,and then let's go full steam ahead with this. 

@qyx268, this will mean that you should set the default of `F_STAR7_MINI` based on the value of `USE_MINI_HALOS` (which you'll have to add). Then for the `pr-feature`, you'll only have to set `USE_MINI_HALOS=True`! I'll still work on extending it though :-)

BradGreig commented:
I think I am in preference of 4 as well. I completely agree that a flag should be set in `FlagOptions` which should be toggled as True/False.  Then the actual value for `F_STAR7_MINI` that is set in `AstroParams` should be used. Throwing up warnings for behaviours as described by @steven-murray.

There needs to be a default value of `F_STAR7_MINI` that gets used if the `USE_X` flag is True but `F_STAR7_MINI` is not set by the user.

catherinewatkinson commented:
I think having a default value if no value is passed is for sure the way to go.

steven-murray commented:
Awesome! Looks like we're all in agreeance! @qyx268  you have a go.

qyx268 commented:
Okay, I've added the flag USE_MINI_HALOS and changed the previous if-condition to using this flag. I've also added warnings when USE_MINI_HALOS=True but USE_MASS_DEPENDENT_ZETA and INHOMO_RECO=False and automatically set the latter two flags to True when minihalos are included. 

I have pr-feature results from a previous comparison when F_STAR7_MINI was set to 0 (because negative values cannot be passed to the variables). This means mini halos have a significant contribution -- WF coupling and X-ray heating happen very early. I'm doing another pr_feature run where,  for the new feature, USE_MINI_HALOS is set to True and consequently F_STAR7_MINI is set to its default value (-2.). I would upload the result once it's done.

**Note that I've modified the script and the default now corresponds to** {"USE_TS_FLUCT": True, "USE_MASS_DEPENDENT_ZETA": True, "INHOMO_RECO": True}. 

[pr_feature_lighcone_2d_brightness_temp.pdf](https://github.com/21cmfast/21cmFAST/files/3971227/pr_feature_lighcone_2d_brightness_temp.pdf)
[pr_feature_history.pdf](https://github.com/21cmfast/21cmFAST/files/3971228/pr_feature_history.pdf)
[pr_feature_power_history.pdf](https://github.com/21cmfast/21cmFAST/files/3971229/pr_feature_power_history.pdf)

steven-murray commented on src/py21cmfast/wrapper.py:
```diff
@@ -322,6 +327,28 @@ def M_MIN_in_Mass(self):
+    @property
+    def USE_MASS_DEPENDENT_ZETA(self):
+        if self.USE_MINI_HALOS and not self._USE_MASS_DEPENDENT_ZETA:
+            print(
```
Let's change `print` to `logger.warning()` and remove the `[WARNING]` from the front. This way it will be logged appropriately.

steven-murray commented on src/py21cmfast/wrapper.py:
```diff
@@ -322,6 +327,28 @@ def M_MIN_in_Mass(self):
+            print(
+                "[WARNING]You have set USE_MINI_HALOS to True but USE_MASS_DEPENDENT_ZETA to False!"
+            )
+            print("         Automatically setting USE_MASS_DEPENDENT_ZETA to True.")
```
Change this `print` to `logger.warning` as well.

steven-murray commented on src/py21cmfast/wrapper.py:
```diff
@@ -322,6 +327,28 @@ def M_MIN_in_Mass(self):
+    @property
+    def INHOMO_RECO(self):
+        if self.USE_MINI_HALOS and not self._INHOMO_RECO:
+            print(
```
Same here for `logger.warning`

qyx268 commented:
okay, I've updated it accordingly. I will wait for the other reviewers' approval before merging the branch into master @andreimesinger @BradGreig @jaehongpark00 .

Also below is the new pr-feature results where DEFAULT adopts {"USE_TS_FLUCT": True, "USE_MASS_DEPENDENT_ZETA": True, "INHOMO_RECO": True} and NEW further includes "USE_MINI_HALOS": True.

[pr_feature_history.pdf](https://github.com/21cmfast/21cmFAST/files/3976502/pr_feature_history.pdf)
[pr_feature_lighcone_2d_brightness_temp.pdf](https://github.com/21cmfast/21cmFAST/files/3976503/pr_feature_lighcone_2d_brightness_temp.pdf)
[pr_feature_power_history.pdf](https://github.com/21cmfast/21cmFAST/files/3976504/pr_feature_power_history.pdf)

steven-murray commented:
@qyx268 I don't think you need everyone's approval. I would wait for at least one more review (who knows the physics better than I), then merge!

BradGreig commented on src/py21cmfast/src/SpinTemperatureBox.c:
```diff
@@ -1123,36 +1403,49 @@ LOG_SUPER_DEBUG("looping over box...");
                         dxion_source_dt_box[box_ct] += (dfcoll_dz_val*(double)del_fcoll_Rct[box_ct]*( (freq_int_ion_tbl_diff[m_xHII_low_box[box_ct]][R_ct])*inverse_val_box[box_ct] + freq_int_ion_tbl[m_xHII_low_box[box_ct]][R_ct] ));
-                    }
-                    else {
-                        dxheat_dt_box[box_ct] += 0.;
```
You have deleted this else statement. Is there a reason why you have done this? This was here to ensure that zero was added if there was some weird numerical issue

BradGreig commented on src/py21cmfast/src/SpinTemperatureBox.c:
```diff
@@ -1434,7 +1766,17 @@ LOG_SUPER_DEBUG("finished loop");
+                J_LW_ave /= (double)HII_TOT_NUM_PIXELS;
+                J_LW_ave_MINI /= (double)HII_TOT_NUM_PIXELS;
+
+                LOG_DEBUG("zp = %e Ts_ave = %e x_e_ave = %e Tk_ave = %e J_alpha_ave = %e(%e) xalpha_ave = %e Xheat_ave = %e(%e) Xion_ave = %e J_LW_ave = %e (%e)",zp,Ts_ave,x_e_ave,Tk_ave,J_alpha_ave,J_alpha_ave_MINI,xalpha_ave,Xheat_ave,Xheat_ave_MINI,Xion_ave,J_LW_ave/1e21,J_LW_ave_MINI/1e21);
```
Its probably best to have the mini halo properties on their own line. It'll be easier to read then.

BradGreig commented on src/py21cmfast/src/SpinTemperatureBox.c:
```diff
@@ -101,9 +123,27 @@ if (LOG_LEVEL >= DEBUG_LEVEL){
+    if (flag_options->USE_MINI_HALOS){
+        log10_Mcrit_LW_unfiltered = (fftwf_complex *) fftwf_malloc(sizeof(fftwf_complex)*HII_KSPACE_NUM_PIXELS);
+        log10_Mcrit_LW_filtered = (fftwf_complex *) fftwf_malloc(sizeof(fftwf_complex)*HII_KSPACE_NUM_PIXELS);
+        log10_Mcrit_LW = (float **) calloc(global_params.NUM_FILTER_STEPS_FOR_Ts,sizeof(float *));
```
These do not appear to be free'd at the end of the file. Please make sure all declared memory is free'd.

steven-murray replied:
It's also worth running a test script through once with `valgrind` to check for these kinds of errors.

BradGreig commented on src/py21cmfast/wrapper.py:
```diff
@@ -1256,6 +1345,9 @@ def ionize_box(
         `redshift` must be given, or `user_params`, `cosmo_params` and `redshift`.
 
+    previous_perturbed_field : :class:`~PerturbField`, optional
+        An perturbed field at higher redshift. This is only used if mini_halo is included.
```
You have added the previous_perturbed_field argument and added it to ionised box.

However, in the run_lightcone and run_coeval classes etc. (and possibly other functions) you haven't added the previous_perturbed_field argument to the calculation.

BradGreig commented on src/py21cmfast/src/ps.c:
```diff
@@ -21,8 +21,12 @@ static gsl_spline *erfc_spline;
-#define NGL_SFR 100
-
+//#define NSFR_high 50 //200
+//#define NSFR_low 50 //250
```
Remove debugging code and alternate values etc. Only include the fiducial values used for the calculations

BradGreig commented on src/py21cmfast/src/ps.c:
```diff
@@ -21,8 +21,12 @@ static gsl_spline *erfc_spline;
+#define NGL_SFR 100 // 100
+#define NMTURN 50//100
+#define LOG10_MTURN_MAX (double) (10)
+#define LOG10_MTURN_MIN (double) (5.-9e-8)
```
Why can't this just be 5.?

JulianBMunoz replied:
Isn't LOG10_MTURN_MAX = 10 super large? It should never be much larger than Matom~7-8.

qyx268 replied:
I've added -9e-8 for some round error. In most cases, critical mass for photodissociation is smaller than 1e8.5, but for UV ionizing,it can reach ~1e9.6.

BradGreig commented on src/py21cmfast/src/ps.c:
```diff
@@ -2040,6 +2584,148 @@ int initialise_SFRD_Conditional_table(int Nfilter, float min_density[], float ma
+
+    ln_10 = log(10);
+
+    Mmin = 1e5;
```
Should be put into global_params

steven-murray commented on src/py21cmfast/src/Constants.h:
```diff
@@ -15,16 +15,8 @@
-
-
-#define R_BUBBLE_MIN (float) (L_FACTOR*1)
-// Minimum radius of an HII region in cMpc.  One can set this to 0, but should be careful with
```
Agreed!

steven-murray commented on src/py21cmfast/src/IonisationBox.c:
```diff
@@ -40,29 +40,45 @@ LOG_DEBUG("redshift=%f, prev_redshift=%f", redshift, prev_redshift);
 
     int overdense_int, overdense_int_boundexceeded;
     int something_finite_or_infinite = 0;
+    int log10_M_MINm_int, log10_M_MINa_int;
```
I heartily agree with @BradGreig on this. Also, "explicit is better than implicit" -- longer variable names are better if they make it more obvious what the variable is about.

JulianBMunoz commented on src/py21cmfast/wrapper.py:
```diff
@@ -338,18 +365,24 @@ class AstroParams(StructWithDefaults):
         but is required by this class to set some default behaviour.
     HII_EFF_FACTOR : float, optional
     F_STAR10 : float, optional
+    F_STAR7_MINI : float, optional
```
Here could go the explanation of why it's 7 vs 10 for F_STAR.

qyx268 replied:
1e10 is a typical value for atomic cooling halos while 1e7 is typical for minihalos. we don't talk about 1e7 atomic cooling halos or 1e10 minihalos.

andreimesinger commented:
> On 07.01.2020., at 16:57, Julian Munoz <notifications@github.com> wrote:
> 
> +            log10_M_MINa_ave /= HII_TOT_NUM_PIXELS;
> +            log10_M_MINm_ave /= HII_TOT_NUM_PIXELS;
> 

just to chime in briefly, as this caught my eye: it is good practice (at least it used to be back in my day) to cast ints and longs to floats or doubles when performing division, so as to avoid integer rounding

qyx268 commented on src/py21cmfast/src/IonisationBox.c:
```diff
@@ -209,9 +268,68 @@ LOG_SUPER_DEBUG("density field calculated");
+        }
+        else{
+            M_MIN = astro_params->M_TURN/50.;
+            M_MINa = astro_params->M_TURN;
```
I will keep the style as other variables where BLABLA is for atomic-cooling halos while BLABLA_MINI is for minihalos. e.g. Mturnover and Mturnover_MINI for this M_MINa and M_MINm

qyx268 commented on src/py21cmfast/src/IonisationBox.c:
```diff
@@ -698,34 +1213,48 @@ LOG_ULTRA_DEBUG("while loop for until RtoM(R)=%f reaches M_MIN=%f", RtoM(R), M_M
             Gamma_R_prefactor = (R*CMperMPC) * SIGMA_HI * global_params.ALPHA_UVB / (global_params.ALPHA_UVB+2.75) * N_b0 * ION_EFF_FACTOR / 1.0e-12;
+            Gamma_R_prefactor_MINI = (R*CMperMPC) * SIGMA_HI * global_params.ALPHA_UVB / (global_params.ALPHA_UVB+2.75) * N_b0 * ION_EFF_FACTOR_MINI / 1.0e-12;
             if(flag_options->PHOTON_CONS) {
                 // Used for recombinations, which means we want to use the original redshift not the adjusted redshift
```
@andreimesinger is it okay if I just automatically turn off photon conservation when minihalos are included? We can come back to this when we need these two features together.

qyx268 commented on src/py21cmfast/src/IonisationBox.c:
```diff
@@ -74,11 +90,14 @@ LOG_DEBUG("redshift=%f, prev_redshift=%f", redshift, prev_redshift);
     }
     else {
         ION_EFF_FACTOR = astro_params->HII_EFF_FACTOR;
+        ION_EFF_FACTOR_MINI = 0.;
```
yes, if USE_MASS_DEPENDENT_ZETA is 0, even Park+18 model is not switched off, we skip all calculations which have mass dependence and go back to the 3-parameter model where Fcoll is adopted.

qyx268 commented on src/py21cmfast/src/21cmFAST.h:
```diff
@@ -45,12 +45,18 @@ struct AstroParams{
     float F_ESC10;
     float ALPHA_ESC;
     float M_TURN;
+    float F_STAR7_MINI;
```
7 means this is normalized over 1e7 Msol. e.g. when F_STAR7_MINI=0.5, fstar for minihalos becomes 10^0.5 * (M/1e7)^ALPHA_STAR.

qyx268 commented on src/py21cmfast/src/IonisationBox.c:
```diff
@@ -209,9 +268,68 @@ LOG_SUPER_DEBUG("density field calculated");
+            }
+            log10_M_MINa_ave /= HII_TOT_NUM_PIXELS;
+            log10_M_MINm_ave /= HII_TOT_NUM_PIXELS;
+            M_MINa      = pow(10., log10_M_MINa_ave);
```
they are used to calculate the mean f_coll e.g. (Nion_General)

qyx268 commented on src/py21cmfast/src/SpinTemperatureBox.c:
```diff
@@ -1123,36 +1403,49 @@ LOG_SUPER_DEBUG("looping over box...");
                 }
 
+                if(ave_fcoll_MINI!=0.) {
+                    ave_fcoll_inv_MINI = 1./ave_fcoll_MINI;
```
this is 1/avg(fcoll), we use ave(fcoll) to scale the mean from PS to SMT.

qyx268 commented on src/py21cmfast/src/IonisationBox.c:
```diff
@@ -244,27 +370,53 @@ LOG_SUPER_DEBUG("sigma table has been initialised");
+                box->mean_f_coll = previous_ionize_box->mean_f_coll + Nion_General(redshift,M_MIN,M_MINa,astro_params->ALPHA_STAR,astro_params->ALPHA_ESC,astro_params->F_STAR10,astro_params->F_ESC10,Mlim_Fstar,Mlim_Fesc) - Nion_General(prev_redshift,M_MIN,M_MINa,astro_params->ALPHA_STAR,astro_params->ALPHA_ESC,astro_params->F_STAR10,astro_params->F_ESC10,Mlim_Fstar,Mlim_Fesc);
+            }
+            if (previous_ionize_box->mean_f_coll_MINI < 1e-15){
+                box->mean_f_coll_MINI = Nion_General_MINI(redshift,M_MIN,M_MINm,Mcrit_atom,astro_params->ALPHA_STAR,astro_params->ALPHA_ESC,astro_params->F_STAR7_MINI,astro_params->F_ESC7_MINI,Mlim_Fstar_MINI,Mlim_Fesc_MINI);
```
sorry, it's been a while. I'm not sure if I follow this question. we calculate mean_f_coll here using ST, then later we use this value rescale the PS-base caculated fcoll to ST.

qyx268 commented on src/py21cmfast/src/SpinTemperatureBox.c:
```diff
@@ -1167,12 +1460,28 @@ LOG_SUPER_DEBUG("looping over box...");
                         dxlya_dt_box[box_ct] *= const_zp_prefactor*prefactor_1 * (1.+delNL0[0][box_ct]*growth_factor_zp);
                         dstarlya_dt_box[box_ct] *= prefactor_2;
+                        if (flag_options->USE_MINI_HALOS){
+                            dstarlyLW_dt_box[box_ct] *= prefactor_2 * (hplank * 1e21);
```
we will be using J21 to calculate Mcrit for photodissociation. This value is in order of 1e-21, therefore, I multiply 1e21 here in case float variables cannot record this small value.

JulianBMunoz replied:
Ah I see, so it's not units, just normalization. Thanks!

qyx268 commented on src/py21cmfast/src/heating_helper_progs.c:
```diff
@@ -836,8 +854,57 @@ double integrate_over_nu(double zp, double local_x_e, double lower_int_limit, in
+
+    fcoll = Nion_z_val[redshift_int_fcollz] + ( zhat - redshift_table_fcollz )*( Nion_z_val[redshift_int_fcollz+1] - Nion_z_val[redshift_int_fcollz] )/(zpp_bin_width);
+
+    log10_Mturn_MINI_int_fcollz = (int)floor( ( log10_Mturn_MINI - LOG10_MTURN_MIN) / LOG10_MTURN_INT);
```
ah, now I understand the question. @andreimesinger do you remember why we have decided to do Fcoll(avg(logM_turn)) instead of avg(Fcoll((logM_turn))) for the ST normalization? I remember we had similar discussions.

qyx268 replied:
At high z, the PDF of logM_turn is very peaky (see the ipython notebook). So these two should be close.

qyx268 commented on src/py21cmfast/src/UsefulFunctions.c:
```diff
@@ -765,3 +765,22 @@ void print_corners_real(float *x, int size){
+    return TtoM(z, 1e4, 0.59);
+}
+
+double molecular_cooling_threshold(float z){
```
ha! I just took this equation (774) from literature. (e.g. Visbal+17).. Taking T=600K and mu=0.59, I got the same normalization for our cosmology. But maybe it's better to use mu=1.22? which then give 0.34 times 3.314e7...

andreimesinger commented:
why would you need to turn off photon conservation??  these two features should be compatible…  in principle, photon conservation should be on by default

> On 09.01.2020., at 03:01, Yuxiang Qin <notifications@github.com> wrote:
> 
> @qyx268 commented on this pull request.
> 
> In src/py21cmfast/src/IonisationBox.c <https://github.com/21cmfast/21cmFAST/pull/37#discussion_r364529670>:
> 
> > @@ -698,34 +1213,48 @@ LOG_ULTRA_DEBUG("while loop for until RtoM(R)=%f reaches M_MIN=%f", RtoM(R), M_M
>  
>              xHI_from_xrays = 1;
>              Gamma_R_prefactor = (R*CMperMPC) * SIGMA_HI * global_params.ALPHA_UVB / (global_params.ALPHA_UVB+2.75) * N_b0 * ION_EFF_FACTOR / 1.0e-12;
> +            Gamma_R_prefactor_MINI = (R*CMperMPC) * SIGMA_HI * global_params.ALPHA_UVB / (global_params.ALPHA_UVB+2.75) * N_b0 * ION_EFF_FACTOR_MINI / 1.0e-12;
>              if(flag_options->PHOTON_CONS) {
>                  // Used for recombinations, which means we want to use the original redshift not the adjusted redshift
> @andreimesinger <https://github.com/andreimesinger> is it okay if I just automatically turn off photon conservation when minihalos are included? We can come back to this when we need these two features together.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/21cmfast/21cmFAST/pull/37?email_source=notifications&email_token=ADH2UAQVZLY6M6TT7KEVNYTQ42AQDA5CNFSM4I5FSDKKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCRD5SDA#discussion_r364529670>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADH2UAUHGW7QVZDBJ5LGMUDQ42AQDANCNFSM4I5FSDKA>.
>

andreimesinger commented:
> On 09.01.2020., at 04:39, Yuxiang Qin <notifications@github.com> wrote:
> 
> @qyx268 commented on this pull request.
> 
> In src/py21cmfast/src/heating_helper_progs.c <https://github.com/21cmfast/21cmFAST/pull/37#discussion_r364547209>:
> 
> > +    float redshift_table_fcollz, log10_Mturn_MINI_table_fcollz;
> +
> +    tauX_params_approx *p = (tauX_params_approx *) params;
> +
> +    drpropdz = C * dtdz(zhat);
> +    n = N_b0 * pow(1+zhat, 3);
> +    nuhat = p->nu_0 * (1+zhat);
> +	log10_Mturn_MINI = p->log10_Mturn_MINI;
> +	LOG10_MTURN_INT = p->LOG10_MTURN_INT;
> +
> +    redshift_int_fcollz = (int)floor( ( zhat - determine_zpp_min )/zpp_bin_width );
> +    redshift_table_fcollz = determine_zpp_min + zpp_bin_width*(float)redshift_int_fcollz;
> +
> +    fcoll = Nion_z_val[redshift_int_fcollz] + ( zhat - redshift_table_fcollz )*( Nion_z_val[redshift_int_fcollz+1] - Nion_z_val[redshift_int_fcollz] )/(zpp_bin_width);
> +
> +    log10_Mturn_MINI_int_fcollz = (int)floor( ( log10_Mturn_MINI - LOG10_MTURN_MIN) / LOG10_MTURN_INT);
> ah, now I understand the question. @andreimesinger <https://github.com/andreimesinger> do you remember why we have decided to do Fcoll(avg(logM_turn)) instead of avg(Fcoll((logM_turn))) for the ST normalization? I remember we had similar discussions.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/21cmfast/21cmFAST/pull/37?email_source=notifications&email_token=ADH2UAXNPQOBJHUQFZWEWTLQ42MAFA5CNFSM4I5FSDKKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCREC3DY#discussion_r364547209>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADH2UAUI66ZUG32OJRUF6R3Q42MAFANCNFSM4I5FSDKA>.
> 


because if you define f_coll only on the cell size, then you are using the conditional mass function in a regime for which it wasn’t really intended: quisi non-linear, and the results are more resolution dependent…  see footnote 4 in Sobacchi and Mesinger (2013)
https://watermark.silverchair.com/stt693.pdf?token=AQECAHi208BE49Ooan9kkhW_Ercy7Dm3ZL_9Cf3qfKAc485ysgAAAoIwggJ-BgkqhkiG9w0BBwagggJvMIICawIBADCCAmQGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMZlzdfd7Jrb4GXcfYAgEQgIICNRsztydw1aAbNZCO1zvyFpn-fy4ofMOzjuEwi_8tjI3qY9_vhtO8zadq7Fl04ietI66YkpaJY-TQdEkp9hvuOTAh_XfjibXSYA1qo4hMfZgbllAdfeuwEVQNZ71ODuSVmXmNvcw-dDfHHZijzPOmQwk-90LDfi0lCtQFQk2kBEP35RdREot5Nt51LuYGOtRpEHjWup6oi8E5G78XygQc6zikDuPKd3pFW7UTeztPHI44Dn9bDCM-UurE5SWzxW4rlF_KXjrakuylA7B7_9jL4xlp2eFennpo6IvyKER6YQjPBx2Q8TVt3eJEjNi4cXedtFJM8jYnAGWaPpARx5pyTiDH_4kAhav6Odn4BRYcgbfHGph7a82p3H2y7iRhZdtcR5J_C2tOSZR1U3B6-2ZtO5V2NOUeEBF10aD22rFjHWRL_ud_-knJZ5RQu6l0vgl7T4RT5uLGLzrAD5lIk71NtsZsabLhr19LdZMFVtljd44QkFbS82GVrBtVdKSwGF55mYjmDhJCKU5B7crD5nRja_fXX4Quh5IJByU6TFGiegEvZK8WXtqv0j1BwkkSxtCvjttIp1FW_uWtPOjiRt7Il6P5rUkEyHQNpS6tzQ_eilaZWiYjpJvv8KILJHagHRgc2DXcSCWHlgirG135r6SSmsY2iRWt7mIuzlNfvXzYtpQY-rYhouQbo9prc4CWKT7aOr2oX45P7zY7d_iqGGUUHCzauX_yfrXkmFD7401aa2D1Es-09H0 <https://watermark.silverchair.com/stt693.pdf?token=AQECAHi208BE49Ooan9kkhW_Ercy7Dm3ZL_9Cf3qfKAc485ysgAAAoIwggJ-BgkqhkiG9w0BBwagggJvMIICawIBADCCAmQGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMZlzdfd7Jrb4GXcfYAgEQgIICNRsztydw1aAbNZCO1zvyFpn-fy4ofMOzjuEwi_8tjI3qY9_vhtO8zadq7Fl04ietI66YkpaJY-TQdEkp9hvuOTAh_XfjibXSYA1qo4hMfZgbllAdfeuwEVQNZ71ODuSVmXmNvcw-dDfHHZijzPOmQwk-90LDfi0lCtQFQk2kBEP35RdREot5Nt51LuYGOtRpEHjWup6oi8E5G78XygQc6zikDuPKd3pFW7UTeztPHI44Dn9bDCM-UurE5SWzxW4rlF_KXjrakuylA7B7_9jL4xlp2eFennpo6IvyKER6YQjPBx2Q8TVt3eJEjNi4cXedtFJM8jYnAGWaPpARx5pyTiDH_4kAhav6Odn4BRYcgbfHGph7a82p3H2y7iRhZdtcR5J_C2tOSZR1U3B6-2ZtO5V2NOUeEBF10aD22rFjHWRL_ud_-knJZ5RQu6l0vgl7T4RT5uLGLzrAD5lIk71NtsZsabLhr19LdZMFVtljd44QkFbS82GVrBtVdKSwGF55mYjmDhJCKU5B7crD5nRja_fXX4Quh5IJByU6TFGiegEvZK8WXtqv0j1BwkkSxtCvjttIp1FW_uWtPOjiRt7Il6P5rUkEyHQNpS6tzQ_eilaZWiYjpJvv8KILJHagHRgc2DXcSCWHlgirG135r6SSmsY2iRWt7mIuzlNfvXzYtpQY-rYhouQbo9prc4CWKT7aOr2oX45P7zY7d_iqGGUUHCzauX_yfrXkmFD7401aa2D1Es-09H0>

this being said, i know Fred Davies is doing just this with an earlier version of the code, and says the ionization topology looks good, perhaps even a bit better than in the default setting.  it mostly just screws up the EoR history, since then the photon conservation should be much worse.  however, now that we have the photon conservation that can correct for this bias, we might want to revisit this at some point.  it would simplify several calculations, since the SFR and associated emissivities can just be defined on the grid, and all filtering performed directly on that grid.

andreimesinger commented:
minihalos would be more relevant in the mostly neutral IGM, so mu=1.2 as default makes a bit more sense…  but i don’t think this matters much if you are just consistent and use only one choice

> On 09.01.2020., at 05:35, Yuxiang Qin <notifications@github.com> wrote:
> 
> @qyx268 commented on this pull request.
> 
> In src/py21cmfast/src/UsefulFunctions.c <https://github.com/21cmfast/21cmFAST/pull/37#discussion_r364556074>:
> 
> > @@ -765,3 +765,22 @@ void print_corners_real(float *x, int size){
>      }
>      printf("\n");
>  }
> +
> +double atomic_cooling_threshold(float z){
> +    return TtoM(z, 1e4, 0.59);
> +}
> +
> +double molecular_cooling_threshold(float z){
> ha! I just took this equation (774) from literature. (e.g. Visbal+17).. Taking T=600K and mu=0.59, I got the same normalization for our cosmology. But maybe it's better to use mu=1.22? which then give 0.34 times 3.314e7...
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/21cmfast/21cmFAST/pull/37?email_source=notifications&email_token=ADH2UAXGV3DAXUFTIJLQGK3Q42SRBA5CNFSM4I5FSDKKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCREFO4Y#discussion_r364556074>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADH2UATSU3EBCWLU7MM5UR3Q42SRBANCNFSM4I5FSDKA>.
>

qyx268 commented:
> minihalos would be more relevant in the mostly neutral IGM, so mu=1.2 as default makes a bit more sense… but i don’t think this matters much if you are just consistent and use only one choice
> […](#)
> On 09.01.2020., at 05:35, Yuxiang Qin ***@***.***> wrote: @qyx268 commented on this pull request. In src/py21cmfast/src/UsefulFunctions.c <[#37 (comment)](https://github.com/21cmfast/21cmFAST/pull/37#discussion_r364556074)>: > @@ -765,3 +765,22 @@ void print_corners_real(float *x, int size){ } printf("\n"); } + +double atomic_cooling_threshold(float z){ + return TtoM(z, 1e4, 0.59); +} + +double molecular_cooling_threshold(float z){ ha! I just took this equation (774) from literature. (e.g. Visbal+17).. Taking T=600K and mu=0.59, I got the same normalization for our cosmology. But maybe it's better to use mu=1.22? which then give 0.34 times 3.314e7... — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <#37?email_source=notifications&email_token=ADH2UAXGV3DAXUFTIJLQGK3Q42SRBA5CNFSM4I5FSDKKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCREFO4Y#discussion_r364556074>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADH2UATSU3EBCWLU7MM5UR3Q42SRBANCNFSM4I5FSDKA>.

I think I would keep using the original one instead of using the TtoM function. TtoM(T=600K, mu=1.22) gives <1e5Msol at z~30. It's not very necessary to go down this low-mass range. Secondly, the fitting function provided originally from Machacek+01 has a floor of 1.25e5Msol

qyx268 commented:
> why would you need to turn off photon conservation?? these two features should be compatible… in principle, photon conservation should be on by default
> […](#)
> On 09.01.2020., at 03:01, Yuxiang Qin ***@***.***> wrote: @qyx268 commented on this pull request. In src/py21cmfast/src/IonisationBox.c <[#37 (comment)](https://github.com/21cmfast/21cmFAST/pull/37#discussion_r364529670)>: > @@ -698,34 +1213,48 @@ LOG_ULTRA_DEBUG("while loop for until RtoM(R)=%f reaches M_MIN=%f", RtoM(R), M_M xHI_from_xrays = 1; Gamma_R_prefactor = (R*CMperMPC) * SIGMA_HI * global_params.ALPHA_UVB / (global_params.ALPHA_UVB+2.75) * N_b0 * ION_EFF_FACTOR / 1.0e-12; + Gamma_R_prefactor_MINI = (R*CMperMPC) * SIGMA_HI * global_params.ALPHA_UVB / (global_params.ALPHA_UVB+2.75) * N_b0 * ION_EFF_FACTOR_MINI / 1.0e-12; if(flag_options->PHOTON_CONS) { // Used for recombinations, which means we want to use the original redshift not the adjusted redshift @andreimesinger <https://github.com/andreimesinger> is it okay if I just automatically turn off photon conservation when minihalos are included? We can come back to this when we need these two features together. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <#37?email_source=notifications&email_token=ADH2UAQVZLY6M6TT7KEVNYTQ42AQDA5CNFSM4I5FSDKKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCRD5SDA#discussion_r364529670>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADH2UAUHGW7QVZDBJ5LGMUDQ42AQDANCNFSM4I5FSDKA>.

okay, I didn't know that photonCon is going to be on by default. I have the concern that for some models, minihalos can do the ionization and keep it ~10% for a while before atomic cooling halos finish the job. For xH<0.3, the photonCon algorithm uses extrapolation and therefore cannot perseve features such as plataeu or even double ionization.

qyx268 commented:
> I'm OK with using the fit from Visbal, but why keep the "molecular_cooling_threshold" function then? It's not used anywhere else right?

well makes sense... I guess I will just remove them...

qyx268 commented:
Okay, let me just summarize a bit... I think I've solved most comments from @BradGreig @JulianBMunoz and @andreimesinger 

I'm just varying parameters and updating the notebook to show the minihalo feature. I can in principle close this PR and put it into master. But I still have some TOs:
1. produce_integration_test_data and test_integration_features. I need to include another line in produce_integration_test_data.py and also test data for turning on minihalos;
2. write doc for minihalos. I need to work on the notebook a bit more and put it into the documentation.
3. enable minihalo with photoCon... apparently, this is what we want? @andreimesinger 

If people are waiting for this branch before merging theirs, I will close this PR and open individual PRs for each of the TOs... @steven-murray what's your opinion?

andreimesinger commented:
ok, if there is a technical issue, you can have it off if minihalos are used for this version. maybe brad can comment...  but this should be changed in the future..

> On 13 Jan 2020, at 06:01, Yuxiang Qin <notifications@github.com> wrote:
> 
> ﻿
> why would you need to turn off photon conservation?? these two features should be compatible… in principle, photon conservation should be on by default
> …
> On 09.01.2020., at 03:01, Yuxiang Qin @.*> wrote: @qyx268 commented on this pull request. In src/py21cmfast/src/IonisationBox.c <#37 (comment)>: > @@ -698,34 +1213,48 @@ LOG_ULTRA_DEBUG("while loop for until RtoM(R)=%f reaches M_MIN=%f", RtoM(R), M_M xHI_from_xrays = 1; Gamma_R_prefactor = (RCMperMPC) * SIGMA_HI * global_params.ALPHA_UVB / (global_params.ALPHA_UVB+2.75) * N_b0 * ION_EFF_FACTOR / 1.0e-12; + Gamma_R_prefactor_MINI = (RCMperMPC) * SIGMA_HI * global_params.ALPHA_UVB / (global_params.ALPHA_UVB+2.75) * N_b0 * ION_EFF_FACTOR_MINI / 1.0e-12; if(flag_options->PHOTON_CONS) { // Used for recombinations, which means we want to use the original redshift not the adjusted redshift @andreimesinger https://github.com/andreimesinger is it okay if I just automatically turn off photon conservation when minihalos are included? We can come back to this when we need these two features together. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <#37?email_source=notifications&email_token=ADH2UAQVZLY6M6TT7KEVNYTQ42AQDA5CNFSM4I5FSDKKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCRD5SDA#discussion_r364529670>, or unsubscribe https://github.com/notifications/unsubscribe-auth/ADH2UAUHGW7QVZDBJ5LGMUDQ42AQDANCNFSM4I5FSDKA.
> 
> okay, I didn't know that photonCon is going to be on by default. I have the concern that for some models, minihalos can do the ionization and keep it ~10% for a while before atomic cooling halos finish the job. For xH<0.3, the photonCon algorithm uses extrapolation and therefore cannot perseve features such as plataeu or even double ionization.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

qyx268 commented:
Okay, let me just summarize a bit... I think I've solved most comments from @BradGreig @JulianBMunoz and @andreimesinger 

I'm just varying parameters and updating the notebook to show the minihalo feature. I can in principle close this PR and put it into master. But I still have some TOs:
1. produce_integration_test_data and test_integration_features. I need to include another line in produce_integration_test_data.py and also test data for turning on minihalos;
2. write doc for minihalos. I need to work on the notebook a bit more and put it into the documentation.
3. enable minihalo with photoCon... apparently, this is what we want? @andreimesinger 

If people are waiting for this branch before merging theirs, I will close this PR and open individual PRs for each of the TOs... @steven-murray what's your opinion?

steven-murray commented:
Thanks @qyx268. I would probably do all of those three things as separate PRs (or maybe one separate PR, if you want). I think the correct way to do this is to make issues (one for each of your points). Then we know we won't forget to fix them.

qyx268 commented:
There was a major bug where previous perturb fields are actually current perturb field. After fixing this bug, I discovered that we need to impose non-negative growth of Fcoll due to slightly evolved density. I've updated the mini-halos python notebook in tutorials. I will close this PR once I've updated the pr-feature results.

steven-murray commented:
@qyx268 -- your last commit seems to indicate this is ready for merging. Do you want a final check-over from anyone, or shall we merge?

Did you make the issues for the three points you mentioned in a comment above?

steven-murray commented:
@qyx268 that's great!

qyx268 commented:
sorry for the delay, but I just found weird results when I increase the box size... 

{"HII_DIM":50, "BOX_LEN": 100}
![image](https://user-images.githubusercontent.com/15994713/72958790-46e21380-3dce-11ea-8f20-6992f61f3eb5.png)

{"HII_DIM":128, "BOX_LEN": 250}
![image](https://user-images.githubusercontent.com/15994713/72958800-4f3a4e80-3dce-11ea-9ca8-5f973af20bdb.png)


You see that they have similar resolution but different sizes. I've always been testing on the small box, the lightcone and global signal and 21cm PS look okay. However, the larger box has this weird boxy structure... 

Do we have the function to output desired LCs? It would be easier for me to check where things go wrong..

steven-murray commented:
Whoa! That's so weird! The lightcone function is `run_lightcone` in `wrapper.py`. Is that what you mean?

qyx268 commented:
these results are from run_lightcone. I was referring to this Issue #39

steven-murray commented:
Ah! I'll get onto that right now -- hopefully will not take very long. Look out for a PR!

qyx268 commented:
{"HII_DIM":50, "BOX_LEN": 250} looks fine...

![image](https://user-images.githubusercontent.com/15994713/73052618-8f233380-3eab-11ea-83af-0d9d4b4947ea.png)

I suspect it has something to do with those interpolation tables that have a large dynamic range

qyx268 commented:
It seems that with {"HII_DIM":128, "BOX_LEN": 250}, not every LC slice has the problem. See these are all 128 slices... looks like there are some ripples in the end...

qyx268 commented:
it's already painful to merge better-lightcone into this branch... I'd rather do it once. I'm gonna try to merge master directly into mini-halo before debugging it...

qyx268 commented:
it looks like the conflict only exists in the wrapper?

BradGreig commented:
@qyx268 I haven't yet had time to merge #71. Does your comment above mean you are advanced in merging #37? (i.e. I should wait). Or, should I still go ahead and do it before you?

qyx268 commented:
I'm trying to fix it this week. If you can wait for a week, please let me merge first. If I don't finish it by this week, go ahead and merge yours. Does this sound okay to you?

qyx268 commented:
yes! finally... plz somebody merges it. 

Here are some results, which are also included in the tutorial.

lightcones
![image](https://user-images.githubusercontent.com/15994713/76708396-bb627000-66f6-11ea-9952-c4ac949c9558.png)

global properties when varying minihalo parameters
![image](https://user-images.githubusercontent.com/15994713/76708428-011f3880-66f7-11ea-9c76-374b9bb224c7.png)

corresponding power spectra
![image](https://user-images.githubusercontent.com/15994713/76708434-0da39100-66f7-11ea-8014-56a683cc8538.png)

steven-murray commented:
@qyx thanks for running the huge mini halos integration test. I'm actually going to make sure that tests skips for now, because it'll be super frustrating to have to wait 45 minutes every time Travis runs tests. It's good to see that it has passed though!
